tests/test_nodes.py

```diff
@@ -560,3 +568,87 @@ def test_allow_objects() -> None:
     iv = IllegalType()
     c.foo = iv
     assert c.foo == iv
+
+
+@pytest.mark.parametrize(  # type: ignore
+    "node, value, expected",
+    [
+        (UnionNode(ref_type=Union[int, str], element_types=[int, str], value=1), 1, 1),
+        (UnionNode(element_types=[int, str], value="str"), "str", "str"),
+        (UnionNode(element_types=[int, str]), "str", "str"),
+        (UnionNode(element_types=[int, str, type(None)]), None, None),
+        (
+            UnionNode(element_types=[DictConfig, str]),
+            DictConfig({"a": 1}),
+            DictConfig({"a": 1}),
+        ),
+        (
+            UnionNode(element_types=[ListConfig, int]),
+            ListConfig([1, 2]),
+            ListConfig([1, 2]),
+        ),
+        (UnionNode(element_types=[List[int], int]), [1, 2], [1, 2]),
+        (UnionNode(element_types=[List[bool], bool]), [True, False], [True, False]),
+        (UnionNode(element_types=[List[float], float]), [3.14, 9.67], [3.14, 9.67]),
+        (UnionNode(element_types=[List[str], str]), ["foo", "foo2"], ["foo", "foo2"]),
+        (
+            UnionNode(element_types=[List[User], str]),
+            [User(name="user1", age=21), User(name="user2", age=45)],
+            [User(name="user1", age=21), User(name="user2", age=45)],
+        ),
+        (
+            UnionNode(element_types=[List[int], int]),
+            1,
+            1,
+        ),
+        (UnionNode(element_types=[Dict[str, int], int]), {"foo": 1}, {"foo": 1}),
+        (
+            UnionNode(element_types=[Dict[str, str], str]),
+            {"foo": "var"},
+            {"foo": "var"},
+        ),
+        (
+            UnionNode(element_types=[Dict[str, bool], bool]),
+            {"foo": True},
+            {"foo": True},
+        ),
+        (
+            UnionNode(element_types=[Dict[str, float], float]),
+            {"foo": 3.14},
+            {"foo": 3.14},
+        ),
+        (
+            UnionNode(element_types=[bool, float], is_optional=True),
+            None,
+            None,
+        ),
+    ],
+)
+def test_valid_value_union_node(node: ValueNode, value: Any, expected: Any) -> None:
+    node._set_value(value)
+    assert node._value() == expected
+    assert isinstance(node, UnionNode)
+
+
+@pytest.mark.parametrize(  # type: ignore
+    "node, value",
+    [
+        (UnionNode(element_types=[int, str], value="str"), User()),
+        (UnionNode(element_types=[int, ListConfig]), [1, 2]),
+        (UnionNode(element_types=[int, ListConfig]), {"foo": "var"}),
+        (UnionNode(element_types=[int, DictConfig]), {"foo": "var"}),
```

Comment:
While trying to understand why assigning a dict was not allowed, I came across the following behavior I find surprising.

When initializing an `Union[DictConfig, int]` with an int, the node is a Union as expected: 

```python
@dataclass
class MyConf1:
    x: Union[DictConfig, int] = 0

c1 = OmegaConf.structured(MyConf1)
print(type(c1._get_node("x")))  # ==> UnionNode (as expected)
```

But when initializing it with a `DictConfig`, it is not a Union anymore:

```python
@dataclass
class MyConf2:
    x: Union[DictConfig, int] = field(default_factory=lambda: DictConfig({"a": 0}))

c2 = OmegaConf.structured(MyConf2)
print(type(c2._get_node("x")))  # DictConfig!
```

Is that working as intended?

Reply:
hmmmm no, I forgot to test DictConfig and ListConfig as types don't change from UnionNode. I'll check it.

Reply:
Check 45f72c0abfdcce6a495463bab845c65611644efd. I added support for this and tests.